File: py_sorts/radix_sort.py
def radix_sort(kv_arr):
    """
    Radix sort starting from the least significant bit. Will return a sorted list as expected (0, 1, 2, 10, 20)
    :return:
    """

    # First explode the key into tuples
    key_digits = _explode_key_seq(kv_arr)

    # Now use bucket sort starting from the LSB iteratively
    for i in reversed(range(key_digits)):
        bucket_sort(kv_arr, 10, i)

    # Now we need to compact the key sequence
    _compact_key_seq(kv_arr)

    return kv_arr


def _compact_key_seq(kv_arr):
    """
    Helper function that compacts the key sequence (used after sorting)

    :param kv_arr:
    :return:
    """

    for i in range(len(kv_arr)):
        key = "".join([str(k) for k in kv_arr[i][0]])
        kv_arr[i][0] = int(key)

    return kv_arr


def _explode_key_seq(kv_arr):
    """
    Helper function that generates the key sequence to feed into the radix sort function

    :param kv_arr:
    :return:
    """

    max_key = max([kv[0] for kv in kv_arr])

    # This is not the most performant solution
    key_digits = len(str(max_key))

    # Replace the key with a tuple containing each digit
    for i in range(len(kv_arr)):
        key_str = str(kv_arr[i][0]).zfill(key_digits)
        kv_arr[i][0] = tuple([int(key_str[i]) for i in range(key_digits)])

    return key_digits


def bucket_sort(kv_arr, n_buckets, key_idx):
    """
    If we have a sequence of (key, value1, value2, ...) tuples where keys are integers in a known range,
    we have a faster algorithm we can use.

    Slightly modified to be able to assist with radix sort by taking a tuple key and sorting by the index of the tuple

    1. For each entry in our sequence, remove from sequence and insert at the end of bucket[key]
    2. Read each bucket sequentially and insert in the original sequence

    :param kv_arr:
    :param n_buckets:
    :param key_idx:
    :return:
    """

    # Instantiate our buckets
    buckets = [[] for i in range(n_buckets)]

    # Build the buckets
    for kv in kv_arr:
        buckets[kv[0][key_idx]].append(kv)

    idx = 0
    for bucket in buckets:
        for kv in bucket:
            kv_arr[idx] = kv
            idx += 1

    return kv_arr

File: c_sorts/radix_sort.c
#include "radix_sort.h"
#include <stdio.h>
#include <limits.h>
#include <math.h>

int *radix_sort(int *kv_arr, size_t n)
{

    // Guard the trivial case
    if (n == 0) { return kv_arr; }

    // Explode the key into a key tuple with the length corresponding to the digits
    // To do this we create an array that holds pointers to the exploded key which we use bucket struct from buckets to represent
    key_bucket *key_arr = malloc(sizeof(key_bucket) * n);

    int key_digits = _explode_key_seq(kv_arr, key_arr, n);

    // Now use bucket sort starting from LSB iteratively
    for (int i = 0; i < key_digits; i++)
    {
        key_arr = bucket_sort(key_arr, n, 10, i);
    }

    // Now we compact the keys down again
    return _compact_key_seq(kv_arr, n, key_arr, key_digits);
}


int *_compact_key_seq(int *kv_arr, size_t n, key_bucket *key_arr, int key_digits)
{
    // For each entry in the sorted key_arr we need to read through
    // And combine the digits
    for (int i = 0; i < n; i++)
    {
        int combined_key = 0;
        for (int j = 0; j < key_digits; j++)
        {
            combined_key += key_arr[i].arr[j] * pow(10, j);
        }

        // Store in kv_arr
        kv_arr[i] = combined_key;
    }

    return kv_arr;
}


int _explode_key_seq(int *kv_arr, key_bucket *key_arr, size_t n)
{

    // First find the maximum number in our array, this will determine the size of the key array required
    int arr_max = -INT_MAX;
    for (int i = 0; i < n; i++)
    {
        if (kv_arr[i] > arr_max)
        {
            arr_max = kv_arr[i];
        }
    }

    // Get number of digits for this max number
    int key_digits = 0;
    while (arr_max != 0)
    {
        arr_max = arr_max / 10;
        key_digits += 1;
    }

    // Now we can malloc our key_buckets
    // Then for each key we store each digit in a separate entry in the array
    for (int i = 0; i < n; i++)
    {
        key_arr[i].arr = malloc(sizeof(int) * key_digits);
        if (!key_arr[i].arr)
        {
            printf("Could not malloc a key array");
            exit(1);
        }

        // Get the current key
        // Store it in the arr in big-endian format
        int cur_key = kv_arr[i];
        for (int j = key_digits - 1; j >= 0; j--)
        {
            int f = pow(10, j);
            if (f > cur_key)
            {
                key_arr[i].arr[j] = 0;

            } else {
                key_arr[i].arr[j] = cur_key / f;
            }
            cur_key = cur_key % f;
        }
    }

    return key_digits;
}


void bucket_append(bucket *bucket, key_bucket kv)
{
    // First create a new node from the values
    _node *node = malloc(sizeof(_node));
    node->value = kv;
    node->next = NULL;

    // Now we insert it into our linked list at the right place
    // We do this in 4 steps:
    // 1. Check if head/tail node is null, if so update to point at the node and done.
    // 2. If not null, get the current tail node
    // 3. Update the next reference in the current tail node to point at the new node
    // 4. Update the tail reference in the bucket to point at the new node
    if ((!bucket->head) & (!bucket->tail))
    {
        bucket->head = node;
        bucket->tail = node;
    }
    else
    {
        _node *tail = bucket->tail;
        tail->next = node;
        bucket->tail = node;
    }
    bucket->length += 1;
}


key_bucket *bucket_sort(key_bucket *key_arr, size_t n, int n_buckets, int key_idx)
{
    // Create our bucket
    bucket buckets[n_buckets];

    // For each bucket go ahead and initialise the struct
    for (int i = 0; i < n_buckets; i++)
    {
        buckets[i].length = 0;
        buckets[i].head = NULL;
        buckets[i].tail = NULL;
    }

    // Now we loop through our array and build the buckets
    for (int i = 0; i < n; i++)
    {
        bucket_append(&buckets[key_arr[i].arr[key_idx]], key_arr[i]);
    }

    // Now we read each bucket in ascending order and concat them all
    // to make the output
    int idx = 0;
    for (int i = 0; i < n_buckets; i++)
    {
        // If head is not null
        if (buckets[i].head)
        {
            // Iteratively traverse the linked list
            _node *node = buckets[i].head;
            while (node)
            {
                key_arr[idx] = node->value;
                node = node->next;
                idx += 1;

            }
        }
    }

    return key_arr;  

}